test/e2e/azure_apiserver_ilb.go

```diff
@@ -0,0 +1,463 @@
+		}
+
+		if len(allNodes.Items) == 0 {
+			return false, fmt.Errorf("no nodes found in the workload cluster")
+		}
+
+		By("Saving all the worker nodes")
+		workerNodes := make(map[string]corev1.Node, 0)
+		for i, node := range allNodes.Items {
+			if strings.Contains(node.Name, input.ClusterName+"-md-0") {
+				workerNodes[node.Name] = allNodes.Items[i]
+			}
+		}
+		if len(workerNodes) != int(input.ExpectedWorkerNodes) {
+			return false, fmt.Errorf("expected number of worker nodes: %d, got: %d", input.ExpectedWorkerNodes, len(workerNodes))
+		}
+
+		By("Saving all the node-debug pods running on the worker nodes")
+		allNodeDebugPods, err := workloadClusterClientSet.CoreV1().Pods("default").List(ctx, metav1.ListOptions{
+			LabelSelector: "app=node-debug",
+		})
+		if err != nil {
+			return false, fmt.Errorf("failed to list node-debug pods in the workload cluster: %v", err)
+		}
+
+		workerDSPods := make(map[string]corev1.Pod, 0)
+		workerDSPodsTestResult := make(map[string]bool, 0)
+		for _, daemonsetPod := range allNodeDebugPods.Items {
+			if _, ok := workerNodes[daemonsetPod.Spec.NodeName]; ok {
+				workerDSPods[daemonsetPod.Name] = daemonsetPod
+				workerDSPodsTestResult[daemonsetPod.Name] = false
+			}
+		}
+		if len(workerDSPods) != int(input.ExpectedWorkerNodes) {
+			return false, fmt.Errorf("expected number of worker node-debug daemonset pods: %d, got: %d", input.ExpectedWorkerNodes, len(workerDSPods))
+		}
+
+		By("Getting the kubeconfig path for the workload cluster")
+		workloadClusterKubeConfigPath := workloadClusterProxy.GetKubeconfigPath()
+		workloadClusterKubeConfig, err := clientcmd.BuildConfigFromFlags("", workloadClusterKubeConfigPath)
+
+		if err != nil {
+			return false, fmt.Errorf("failed to build workload cluster kubeconfig from flags: %v", err)
+		}
+
+		fmt.Fprintf(GinkgoWriter, "Number of node debug pods deployed on worker nodes: %v\n", len(workerDSPods))
+		for _, nodeDebugPod := range workerDSPods {
+			fmt.Fprintf(GinkgoWriter, "node-debug pod %v is deployed on node %v\n", nodeDebugPod.Name, nodeDebugPod.Spec.NodeName)
+
+			By("Checking the status of the node-debug pod")
+			switch nodeDebugPod.Status.Phase {
+			case corev1.PodPending:
+				fmt.Fprintf(GinkgoWriter, "Pod %s is in Pending phase. Retrying\n", nodeDebugPod.Name)
+				return false /* retry */, nil
+			case corev1.PodRunning:
+				fmt.Fprintf(GinkgoWriter, "Pod %s is in Running phase. Proceeding\n", nodeDebugPod.Name)
+			default:
+				return false, fmt.Errorf("node-debug pod %s is in an unexpected phase: %v", nodeDebugPod.Name, nodeDebugPod.Status.Phase)
+			}
+
+			listOfCommands := map[string][]string{
+				"Hello from node-debug pod": {"sh", "-c", "echo \"Hello from node-debug pod\""},
+				apiServerILBPrivateIP:       {"sh", "-c", "test -f /host/etc && cat /host/etc || echo 'File not found'"}, // /etc/host is mounted as /host/etc/hosts in the node-debug pod
+			}
+			testResult := map[string]bool{
+				"Hello from node-debug pod": false,
```

Comment:
nit: Might be good to save "Hello from node-debug pod" to a variable since it's used in multiple places

Reply:
Good idea, will do.

Reply:
Added via f875002db330906ad0735b66246ef348a78f0857